tests/lib/qos/src/qos_test.c

```diff
@@ -0,0 +1,343 @@
+
+/* Function used to verify internal context variables. */
+static void ctx_verify(struct ctx *expected)
+{
+	TEST_ASSERT_EQUAL(ctx.app_evt_handler,
+			  expected->app_evt_handler);
+	TEST_ASSERT_EQUAL(ctx.timeout_handler_work.work.handler,
+			  expected->timeout_handler_work.work.handler);
+	TEST_ASSERT_EQUAL(ctx.pending_list.head, expected->pending_list.head);
+	TEST_ASSERT_EQUAL(ctx.pending_list.tail, expected->pending_list.tail);
+	TEST_ASSERT_EQUAL(ctx.initialized, expected->initialized);
+	TEST_ASSERT_EQUAL(ctx.message_id_next, expected->message_id_next);
+}
+
+static void dut_event_handler(const struct qos_evt *evt)
+{
+	switch (evt->type) {
+	case QOS_EVT_MESSAGE_NEW:
+	case QOS_EVT_MESSAGE_TIMER_EXPIRED:
+		/* Fall through. */
+	case QOS_EVT_MESSAGE_REMOVED_FROM_LIST:
+		callback_count++;
+		break;
+	default:
+		TEST_FAIL();
+		break;
+	}
+}
+
+static void dut_event_handler_expired(const struct qos_evt *evt)
+{
+	if (evt->type == QOS_EVT_MESSAGE_TIMER_EXPIRED) {
+		callback_count++;
+		return;
+	}
+
+	TEST_FAIL();
+}
+
+static void dut_event_handler_removed(const struct qos_evt *evt)
+{
+	if (evt->type == QOS_EVT_MESSAGE_REMOVED_FROM_LIST) {
+		callback_count++;
+		return;
+	}
+
+	TEST_FAIL();
+}
+
+static void init(void)
+{
+	struct ctx expected =  {
+		.app_evt_handler = NULL,
+		.pending_list.head = NULL,
+		.pending_list.tail = NULL,
+		.initialized = false,
+		.timeout_handler_work.work.handler = NULL,
+		.message_id_next = QOS_MESSAGE_ID_BASE
+	};
+
+	/* Verify internal context structure */
+	ctx_verify(&expected);
+
+	/* Test invalid handler argument */
+	TEST_ASSERT_EQUAL(-EINVAL, qos_init(NULL));
+
+	/* Successful initialization */
+	TEST_ASSERT_FALSE(qos_init(dut_event_handler));
+
+	/* Verify that internal variables has been initialized. */
+	expected.app_evt_handler = &dut_event_handler;
+	expected.initialized = true;
+	expected.timeout_handler_work.work.handler = &timeout_handler_work_fn;
+
+	ctx_verify(&expected);
+
+	/* Test subsequent calls */
+	TEST_ASSERT_EQUAL(-EALREADY, qos_init(dut_event_handler));
+}
+
+void setUp(void)
+{
+	ctx.message_id_next = QOS_MESSAGE_ID_BASE;
+	init();
+}
+
+void tearDown(void)
+{
+	k_work_cancel_delayable(&ctx.timeout_handler_work);
+	memset(&ctx, 0, sizeof(struct ctx));
+	callback_count = 0;
+}
+
+void test_message_add(void)
+{
+	struct qos_metadata *node = NULL, *next_node = NULL;
+	uint8_t count = 0;
+	struct qos_data message = {
+		.heap_allocated = true,
+		.data.buf = var,
+		.data.len = DUMMY_SIZE,
+		.id = qos_message_id_get_next(),
+		.type = MESSAGE_1,
+		.flags = QOS_FLAG_RELIABILITY_ACK_DISABLED
+	};
+
+	/* Test for invalid argument. */
+	TEST_ASSERT_EQUAL(-EINVAL, qos_message_add(NULL));
+
+	/* Add single message. */
+	TEST_ASSERT_FALSE(qos_message_add(&message));
+
+	/* Expect the QOS_EVT_MESSAGE_NEW and QOS_EVT_MESSAGE_REMOVED_FROM_LIST to be called once
+	 * for the message flagged with QOS_FLAG_RELIABILITY_ACK_DISABLED.
+	 */
+	TEST_ASSERT_EQUAL(2, callback_count);
```

Comment:
this doesn't actually check that you got one of each message type instead of 2 of one and none of the other, or that they were received in the correct order

Reply:
Yeah, I'm aware. I didn't bother checking for that in this initial implementation.